src/core/model_config.proto

```diff
@@ -1104,7 +1104,36 @@ message ModelSequenceBatching
   //@@     model instance over the lifetime of the sequence. This
   //@@     is the default strategy.
   //@@
-  message StrategyDirect {}
+  message StrategyDirect {
+    //@@    .. cpp:var:: int32 preferred_batch_size (repeated)
+    //@@
+    //@@       Preferred batch sizes for batching of candidate
+    //@@       sequences. If a batch of one of these sizes can be formed
+    //@@       it will be executed immediately. If not specified a
+    //@@       preferred batch size will be chosen automatically
+    //@@       based on model and GPU characteristics.
+    //@@
+    repeated int32 preferred_batch_size = 1;
+
+    //@@    .. cpp:var:: float minimum_slot_utilization
+    //@@
+    //@@       The rate of slot utilization that the selected preferred batch
+    //@@       size must satisfy, otherwise, the largest possible batch will
+    //@@       be executed. The utilization is calculated based on the number of
+    //@@       ready requests in the selected preferred batch. The utilization
+    //@@       is range from 0 to 1 where 1 represents full utilization.
+    //@@       Default is 0.
+    //@@
+    float minimum_slot_utilization = 2;
```

Comment:
@deadeyegoodwin I added this field because direct sequence batcher reserves batch slot for individual sequence, so there may be the case where: preferred batch size is 4 and the slots are assigned to: null, null, null, seq 1, seq 2, seq 3, user can use this field to let batcher send a larger batch while "null, null, null, seq 1" forms a preferred batch.

Do you mind to provide feedback on how you feel about the preferred batch size handling here.

Reply:
I don't think we want preferred_batch_size at all because of the "null" slot issue that you mention. I think the delay along with minimum_slot_utilization is sufficient. 'minimum_slot_utilization' should be "The minimum slot utilization that must be satisfied to execute the batch before 'max_queue_delay_microseconds' expires. For example, a value of 0.5 indicates that the batch should be executed as soon as 50% or more of the slots are ready even if the max_queue_delay_microseconds timeout has not expired. The default is 1.0, indicating that a batch will not execute before max_queue_delay_microseconds timeout expires unless every batch slot is ready.

Reply:
I mention the default to be 0.0 instead of 1.0 because in protobuf 0 will be the default value if not specified, we shouldn't replace 0 to 1 as 0 is also an acceptable value. But you can also argues that 0 doesn't make the delay meaningful as all batches will be executed immediately, so I am also okay with changing the description to say 0 will be overwritten to 1.0

Reply:
Perhaps just add one final sentence. "max_queue_delay_microseconds will be ignored unless minimum_slot_utilization is set to a non-zero value".
Does that make sense?